### PR TITLE
Performance improvements and cleanup

### PR DESCRIPTION
**Pull request by SimonHeybrock**
In the process of preparations for supporting "out" arguments (see #212) a number of improvements and cleanups were required (otherwise we would probably not see the benefit of out args). Since more changes than anticipated have accumulated, I am opening a separate PR. Out-args will follow later.

Improvements/changes:

- Benchmarks for `neutron::convert`.
- Benchmarks for `transform`.
- Implement changes to `transform` which have previously been made for `transform_in_place`. This gives a 100x speedup by avoiding the slow `ViewIndex::operator[]`.
- Some math rewrites in and related to unit conversion for various speedups. In particular, we use in-place operations where possible and avoid some temporaries. There is certainly room some for more improvements here. One item on the todo list is to avoid initializing outputs in `transform`, which I will tackle in a later PR.
- Remove support for count-densities in `neutron::convert`. While this is convenient, it is not critical and comes with a number of drawbacks:
  - More complicated unit testing.
  - More complicated code. This is of particular importance since the introduction of out args will complicate the existing code, so simplifying things beforehand seems the right way.
  - It goes against the principle of "one function/algorithm for one task" which I am trying to adopt in scipp. My personal view is that a good number of issues in Mantid is due to adapting the opposite principle (algorithms try to handle all possible inputs automatically).

### CODE REVIEW

File: benchmark/CMakeLists.txt [hunk 1]
```diff
@@ -15,6 +15,15 @@ target_link_libraries(histogram_benchmark
                       scipp-core
                       scipp_test_helpers)
 
+add_executable(neutron_convert_benchmark EXCLUDE_FROM_ALL neutron_convert_benchmark.cpp)
+add_dependencies(all-benchmarks neutron_convert_benchmark)
```
Comment (OwenArnold) [inline benchmark/CMakeLists.txt]: Is there a good reason for not having a single executable for the benchmarks?

File: core/include/scipp/core/transform.h [hunk 2]
```diff
@@ -51,6 +52,11 @@ template <class T> struct ValuesAndVariances {
   ValuesAndVariances(T &val, T &var) : values(val), variances(var) {
     expect::sizeMatches(values, variances);
   }
+  ValuesAndVariances &operator=(std::pair<T, T> &&data) {
```
Comment (OwenArnold) [inline core/include/scipp/core/transform.h]: Where exactly does move assignment help?

File: benchmark/CMakeLists.txt (hunk 1 above)
Comment (SimonHeybrock) [reply benchmark/CMakeLists.txt]: I don't know, might be worth looking into.

File: core/include/scipp/core/transform.h (hunk 2 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/transform.h]: It is required to make the assignment in https://github.com/scipp/scipp/blob/2e156538bddbadb09016834430bd1eb8292ebbd0/core/include/scipp/core/transform.h#L227-L230 work. The key is not the assignment from move, but the transition from `pair` to `ValuesAndVariances`.

**Pull request merged**